code/game/machinery/doors/firedoor.dm

```diff
@@ -427,42 +433,77 @@
 /obj/machinery/door/firedoor/do_animate(animation)
 	switch(animation)
 		if("opening")
-			flick("door_opening", src)
+			flick("opening", src)
 			playsound(src, 'sound/machines/firelockopen.ogg', 37, 1)
 		if("closing")
 			playsound(src, 'sound/machines/firelockclose.ogg', 37, 1)
-			flick("door_closing", src)
+			flick("closing", src)
 	return
 
 
 /obj/machinery/door/firedoor/update_icon()
-	cut_overlays()
-	var/list/overlays_to_add = list()
+	var/icon/lights_overlay
+	var/icon/panel_overlay
+	var/icon/weld_overlay
+
+	for (var/cardinal in GLOB.cardinal) //No list copy please good sir
+		var/turf/step_turf = get_step(src, cardinal)
+		for(var/atom/thing as anything in step_turf)
+			if(thing.type in airlock_typecache)
+				switch(cardinal)
+					if(EAST)
+						setDir(SOUTH)
+					if(WEST)
+						setDir(SOUTH)
+					if(NORTH)
+						setDir(WEST)
+					if(SOUTH)
+						setDir(WEST)
+		if (step_turf.density == TRUE)
+			switch(cardinal)
+				if(EAST)
+					setDir(SOUTH)
+				if(WEST)
+					setDir(SOUTH)
+				if(NORTH)
+					setDir(WEST)
+				if(SOUTH)
+					setDir(WEST)
+
+	overlays.Cut()
+	set_light(0)
+	var/do_set_light = FALSE
 
 	if(density)
-		icon_state = "door_closed"
-		if(prying)
-			icon_state = "prying_closed"
-		if(hatch_open)
-			overlays_to_add += "hatch"
-		if(blocked)
-			overlays_to_add += "welded"
+
+		icon_state = "closed"
+		if(panel_open)
+			overlays = panel_overlay
 		if(pdiff_alert)
-			overlays_to_add += "palert"
+			lights_overlay += "palert"
+			do_set_light = TRUE
 		if(dir_alerts)
 			for(var/d=1;d<=4;d++)
 				var/cdir = GLOB.cardinal[d]
 				for(var/i=1;i<=ALERT_STATES.len;i++)
-					if(dir_alerts[d] & (1<<(i-1)))
-						overlays_to_add += new/icon(icon,"alert_[ALERT_STATES[i]]", dir=cdir)
+					if(dir_alerts[d] & BITFLAG(i-1))
+						overlays += new/icon(icon,"alert_[ALERT_STATES[i]]", dir=cdir)
```

Comment:
i know the last guys did it but this 100% should not be making new icons every update_icon.
furthermore all of this doesn't use SSoverlays.